### DISCUSSION TITLE
Query for encrypted values

### DISCUSSION THREAD

**Discussion by emiteze** [pkasila/gorm-crypto#14] [Q&A]
Whats up :)

I got to your libraby by your medium post (which is very good, by the way) and I started using your library in my personal project as I start to learn more about golang and gorm itself. So I decided to deal with encrypted values and started using your lib, saving and reading records according to your documentation and everything worked fine.

I was thinking in a case where I want to select specific records querying for an encrypted field, then I ran into a questioning that I don't know exactly if this could be a feature request or if its just a gorm architecture problem. 

Trying to do something like this gives me 0 records probably because of aes algorithm that outputs a different byte array everytime I encrypt the same data:

```golang
res := userRepository.db.Model(models.User{}).
	Where("phone_number = ?", gormcrypto.EncryptedValue{Raw: 'phonenumber'}).
	Find(&user)
```

So probably the best way to do this is not encrypting my phone number to use on my query, but decrypting the db value instead, which would me give something like this in sql:

```sql
select * from users where pgp_sym_decrypt(phone_number::bytea, 'encryption-key', 'cipher-algo=aes256') = 'phonenumber';
```

So, I know how to do that in sql but I was avoiding the raw query approach of gorm for this scenario. Do you have any idea of how can I query for encrypted values? Since the where clause must be decrypted and your decrypt method probably is only called when gorm is deserializing data from db after the query has been ran, I really don't know if this could be a feature request.

**Comment by pkasila** [comment]
I see your point. I understand why you are trying to avoid the raw query approach, but the problem is that `EncryptedValue`, as well as `SignedValue`, uses `sql.Scanner` and `driver.Valuer` (GoLang's `database/sql` and `database/sql/driver`) functions to encrypt and decrypt the value on the client, so if you want to query data from the database, then you need to use `pgcrypto` functionality directly using the raw queries with `pgp_sym_decrypt` or `pgp_sym_decrypt_bytea`.

So, it's clearly the GORM architecture problem. I had the same issue when I was developing the library, so, I've decided to use the RDBMS-specific tools in the raw queries to decrypt the data to query it because there is no unified way to do so across all RDBMS and their extensions and SQL dialects.

**Why is it happening?**

Actually, it's happening because every time you encrypt something, then the `rand.Reader` is used to generate IV, nonce, etc to generate unique bytes sequence each time, so in case of the data breach, for example, no one won't see that people have the same data in the database fields.
